### PR TITLE
[Stacked] Dialogue add preliminary binary support

### PR DESCRIPTION
**Pull request by ferozco**
## Before this PR
Dialogue would barf on any binary parameters

## After this PR
We support binary parameters

## Possible downsides?
N/A
There is still some work to support `optional<binary>` response which currently returns `Optional<ByteBuffer>`

### CODE REVIEW

File: conjure-java-core/src/integrationInput/java/com/palantir/product/AsyncEteBinaryService.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import javax.annotation.Generated;
+
+@Generated("com.palantir.conjure.java.services.dialogue.DialogueServiceGenerator")
+public interface AsyncEteBinaryService {
+    ListenableFuture<InputStream> postBinary(AuthHeader authHeader, BinaryRequestBody body);
+
+    ListenableFuture<Optional<ByteBuffer>> getOptionalBinaryPresent(AuthHeader authHeader);
```
Comment (iamdanfox) [inline conjure-java-core/src/integrationInput/java/com/palantir/product/AsyncEteBinaryService.java:15]: I know useImmutableBytes was an opt-in thing before, can we just start dialoge on com.palantir.conjure.java.lib.Bytes straight away?

Comment (carterkozak) [reply conjure-java-core/src/integrationInput/java/com/palantir/product/AsyncEteBinaryService.java:15]: This should use a streamed binary type, not bytebuffer or bytes, so I’m not sure there’s a legitimate case for forcing useImmutableBytes here

File: conjure-java-core/src/integrationInput/java/com/palantir/product/DialogueEteBinaryService.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            public InputStream getBinaryFailure(AuthHeader authHeader, int numBytes) {
+                return RemoteExceptions.getUnchecked(
+                        delegate.getBinaryFailure(authHeader, numBytes));
+            }
+        };
+    }
+
+    /** Creates an asynchronous/non-blocking client for a EteBinaryService service. */
+    public static AsyncEteBinaryService async(Channel channel, ConjureRuntime runtime) {
+        return new AsyncEteBinaryService() {
```
Comment (iamdanfox) [inline conjure-java-core/src/integrationInput/java/com/palantir/product/DialogueEteBinaryService.java:192]: stacktraces are gonna be so sad with endless anonymous classes 🤷‍♂️

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java [hunk 3]
```diff
@@ -160,82 +175,37 @@ private MethodSpec asyncClientImpl(ClassName serviceClassName, EndpointDefinitio
...
-    private void addParameter(CodeBlock.Builder request, String endpointName, ArgumentDefinition param) {
-        param.getParamType().accept(new ParameterType.Visitor<CodeBlock.Builder>() {
+    private CodeBlock generateParam(String endpointName, ArgumentDefinition param) {
+        return param.getParamType().accept(new ParameterType.Visitor<CodeBlock>() {
             @Override
-            public CodeBlock.Builder visitBody(BodyParameterType value) {
-                return request.add(
+            public CodeBlock visitBody(BodyParameterType value) {
+                if (param.getType().accept(TypeVisitor.IS_BINARY)) {
+                    return CodeBlock.of("$L.body(runtime.bodySerDe().serialize($L));", "_request", param.getArgName());
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java]: can we either inline the _request or pull it to a constant?

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java [hunk 4]
```diff
@@ -160,82 +175,37 @@ private MethodSpec asyncClientImpl(ClassName serviceClassName, EndpointDefinitio
...
 
-    private void addParameter(CodeBlock.Builder request, String endpointName, ArgumentDefinition param) {
-        param.getParamType().accept(new ParameterType.Visitor<CodeBlock.Builder>() {
+    private CodeBlock generateParam(String endpointName, ArgumentDefinition param) {
+        return param.getParamType().accept(new ParameterType.Visitor<CodeBlock>() {
             @Override
-            public CodeBlock.Builder visitBody(BodyParameterType value) {
-                return request.add(
+            public CodeBlock visitBody(BodyParameterType value) {
+                if (param.getType().accept(TypeVisitor.IS_BINARY)) {
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java:195]: in our glorious new dialogue future, maybe we can even support aliases of binary ?!?!

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java [hunk 5]
```diff
@@ -40,7 +41,6 @@
 import com.squareup.javapoet.ParameterizedTypeName;
 import com.squareup.javapoet.TypeName;
 import com.squareup.javapoet.TypeSpec;
-import io.undertow.util.PathTemplate;
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java:43]: 😂

File: conjure-java-core/src/main/java/com/palantir/conjure/java/visitor/ClassVisitor.java [hunk 6]
```diff
@@ -14,6 +14,7 @@
 import com.palantir.conjure.spec.PrimitiveType;
 import com.palantir.conjure.spec.SetType;
 import com.palantir.conjure.spec.TypeDefinition;
+import com.palantir.dialogue.BinaryRequestBody;
 import com.squareup.javapoet.TypeName;
 import java.io.InputStream;
 import java.util.List;
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/visitor/ClassVisitor.java:32]: I would love to change the name of `ClassVisitor`

File: conjure-java-core/src/main/java/com/palantir/conjure/java/visitor/ClassVisitor.java [hunk 7]
```diff
@@ -39,10 +40,11 @@ public TypeName visitPrimitive(PrimitiveType value) {
             return delegate.visitPrimitive(value);
         }
 
+        // TODO(forozco): handle optional binary response
```
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/visitor/ClassVisitor.java:55]: future work might also include `alias<binary>`

File: conjure-java-core/src/integrationInput/java/com/palantir/product/AsyncEteBinaryService.java (hunk 1 above)
Comment (iamdanfox) [reply conjure-java-core/src/integrationInput/java/com/palantir/product/AsyncEteBinaryService.java:15]: Ah yep good point

**Pull request merged**